odoo/models.py

```diff
@@ -2437,27 +2438,57 @@ def _inherits_join_add(self, current_model, parent_model_name, query):
         return parent_alias
 
     @api.model
-    def _inherits_join_calc(self, alias, fname, query):
+    def _inherits_join_calc(self, alias, fname, query, allow_related=False):
         """
         Adds missing table select and join clause(s) to ``query`` for reaching
-        the field coming from an '_inherits' parent table (no duplicates).
+        the field coming from an '_inherits' parent table (no duplicates)
+        or from related.
 
         :param alias: name of the initial SQL alias
         :param fname: name of inherited field to reach
         :param query: query object on which the JOIN should be added
+        :param allow_related: whether we make joins for explicit
+                              (i.e. not inherited) related non-stored field
         :return: qualified name of field, to be used in SELECT clause
         """
         # INVARIANT: alias is the SQL alias of model._table in query
         model, field = self, self._fields[fname]
-        while field.inherited:
-            # retrieve the parent model where field is inherited from
-            parent_model = self.env[field.related_field.model_name]
-            parent_fname = field.related[0]
-            # JOIN parent_model._table AS parent_alias ON alias.parent_fname = parent_alias.id
-            parent_alias = query.left_join(
-                alias, parent_fname, parent_model._table, 'id', parent_fname,
-            )
-            model, alias, field = parent_model, parent_alias, field.related_field
+        def process_inherited(model, alias, field):
+            while field.inherited:
+                # retrieve the parent model where field is inherited from
+                parent_model = self.env[field.related_field.model_name]
+                parent_fname = field.related[0]
+                # JOIN parent_model._table AS parent_alias ON alias.parent_fname = parent_alias.id
+                parent_alias = query.left_join(
+                    alias, parent_fname, parent_model._table, 'id', parent_fname,
+                )
+                model, alias, field = parent_model, parent_alias, field.related_field
+            return model, alias, field
+
+
+        model, alias, field = process_inherited(model, alias, field)
+        def process_related(model, alias, field):
+            related_field = field.related_field
+            for join_fname in field.related[:-1]:
+                field = model._fields[join_fname]
+                if field.inherited:
+                    model, alias, field = process_inherited(model, alias, field)
+                if field.related and not field.store:
+                    assert field.compute_sudo or field.auto_join, "Non-stored related field must be a chain of fields with compute_sudo=True or auto_join=True"
```

Comment:
Not all conditions should be on `field`.
In the case of `country_code`, the condition should be `country_code.compute_sudo or `country_id.auto_join`.